File: backend/Modules/Evaluacion/EvaluacionService.cs
using CodeMX.Api.Modules.Retos;

namespace CodeMX.Api.Modules.Evaluacion;

/// <summary>
/// Contexto del patrón Strategy. Obtiene los casos de prueba del reto (vía IRetosApi) y
/// delega la ejecución a una estrategia creada por el Factory Method: intenta primero la
/// estrategia remota (Servicio Python) y, si falla, recurre a la estrategia local.
/// </summary>
public class EvaluacionService : IEvaluacionApi
{
    private readonly IEvaluadorStrategyFactory _factory;
    private readonly IRetosApi _retos;

    public EvaluacionService(IEvaluadorStrategyFactory factory, IRetosApi retos)
    {
        _factory = factory;
        _retos = retos;
    }

    public async Task<ResultadoEvaluacion> EvaluarAsync(SolicitudEvaluacion solicitud)
    {
        var casos = await _retos.ObtenerCasosAsync(solicitud.RetoId);

        try
        {
            // Estrategia preferida: ejecución remota en el Servicio Python.
            return await _factory.Crear(TipoEvaluacion.Remota).EjecutarAsync(solicitud, casos);
        }
        catch
        {
            // Respaldo: ejecución local mientras el Servicio Python no esté disponible.
            return await _factory.Crear(TipoEvaluacion.Local).EjecutarAsync(solicitud, casos);
        }
    }
}

File: backend-java/src/main/java/mx/codemx/modules/evaluacion/EvaluacionService.java
package mx.codemx.modules.evaluacion;

import java.util.List;
import mx.codemx.modules.retos.CasoPrueba;
import mx.codemx.modules.retos.RetosApi;
import org.springframework.stereotype.Service;

@Service
public class EvaluacionService implements EvaluacionApi {

    private final EvaluadorStrategyFactory factory;
    private final RetosApi retos;

    public EvaluacionService(EvaluadorStrategyFactory factory, RetosApi retos) {
        this.factory = factory;
        this.retos = retos;
    }

    @Override
    public ResultadoEvaluacion evaluar(SolicitudEvaluacion solicitud) {
        List<CasoPrueba> casos = retos.obtenerCasos(solicitud.retoId());

        try {
            return factory.crear(TipoEvaluacion.REMOTA).ejecutar(solicitud, casos);
        } catch (Exception e) {
            return factory.crear(TipoEvaluacion.LOCAL).ejecutar(solicitud, casos);
        }
    }
}
